File: composite/composite.py
#!/usr/bin/env python3

from abc import ABCMeta, abstractmethod
from typing import List


class Entry(metaclass=ABCMeta):
    @abstractmethod
    def get_name(self):
        pass

    @abstractmethod
    def get_size(self):
        pass

    def add(self, entry: "Entry"):
        raise FileTreatmentException

    def print_list(self):
        return self._print_list()

    @abstractmethod
    def _print_list(self, prefix: str):
        pass

    def to_string(self) -> str:
        return f"{self.get_name()} ({self.get_size()})"


class FileTreatmentException(Exception):
    def __init__(self, *args, **kwargs):
        self.message = "FileTreatmentException"


class File(Entry):
    def __init__(self, name: str, size: int):
        self.name: str = name
        self.size: int = size

    def get_name(self) -> str:
        return self.name

    def get_size(self) -> int:
        return self.size

    def _print_list(self, prefix: str = ""):
        return f"{prefix}/{self.to_string()}\n"


class Directory(Entry):
    def __init__(self, name: str):
        self.name: str = name
        self.directory: List["Entry"] = []

    def get_name(self) -> str:
        return self.name

    def get_size(self) -> int:
        size = 0
        for d in self.directory:
            size += d.get_size()
        return size

    def add(self, entry: "Entry"):
        self.directory.append(entry)

    def _print_list(self, prefix: str = ""):
        out = f"{prefix}/{self.to_string()}\n"
        for entry in self.directory:
            out += entry._print_list(f"{prefix}/{self.name}")
        return out

File: composite/composite.go
package composite

import "strconv"

type entry interface {
	getName() string
	getSize() int
	PrintList(prefix string) string
	Add(entry entry)
}

type baseEntry struct {
	entry
	name string
}

func (be *baseEntry) getName() string {
	return be.name
}

func (be *baseEntry) toString(e entry) string {
	return e.getName() + " (" + strconv.Itoa(e.getSize()) + ")" + "\n"
}

type file struct {
	*baseEntry
	size int
}

func (f *file) getSize() int {
	return f.size
}

func (f *file) PrintList(prefix string) string {
	return prefix + "/" + f.toString(f)
}

type directory struct {
	*baseEntry
	directory []entry
}

func (d *directory) getSize() int {
	size := 0
	for _, entry := range d.directory {
		size += entry.getSize()
	}
	return size
}

func (d *directory) PrintList(prefix string) string {
	list := prefix + "/" + d.toString(d)
	for _, entry := range d.directory {
		list += entry.PrintList(prefix + "/" + d.getName())
	}
	return list
}

func (d *directory) Add(entry entry) {
	d.directory = append(d.directory, entry)
}

func NewFile(name string, size int) *file {
	return &file{
		baseEntry: &baseEntry{name: name},
		size:      size,
	}
}

func NewDirectory(name string) *directory {
	return &directory{baseEntry: &baseEntry{name: name}}
}
